Multihost HLO Runner: Add CLI Option for Output Mode

Zoey-Cheng opened the pull request:
This PR adds a new CLI option `--output_mode` to **multihost_hlo_runner**, making the output mode configurable for the running function.

Adds an optional `--output_mode` argument to `multihost_hlo_runner`.
Supports three output modes: `return_outputs`, `not_return_outputs`, and `return_device_0_outputs`.
Default is `return_outputs`. Invalid values trigger a warning and error message.

**What it Change**

The CLI arg is registeredin hlo_runner_main.cc with a notification. It call existing parsing logic from functional_hlo_runner.h (line 173).

**Self Test**
Try arg return_outputs, not_return_outputs, return_device_0_outputs; empty; error input
`sample.hlo`

```
HloModule test

ENTRY main {
    p0 = f32[8,8] parameter(0)
    p1 = f32[8,8] parameter(1)
    ROOT add = f32[8,8] add(p0,p1)
}

```


Running

```
bazel run xla/tools/multihost_hlo_runner/hlo_runner_main -- sample.hlo --log_output=true --output_mode=return_device_0_outputs
```

reedwm reviewed (changes requested):
What's the motivation for the change? Why not always return all outputs?

Also what does it mean to "return outputs" or to "not return outputs"? The HLO module must compute all outputs in any case. What are the outputs returned from?

reedwm commented:
That makes sense. @Zoey-Cheng, in the flag description, can you give details on what the flag does and the motivation? E.g. "Specify whether outputs are returned or not. Possible values: return_outputs, not_return_outputs, return_device0_outputs. If outputs are not returned, the outputs are still computed but a potentially-slow device-to-host copy is skipped."

Zoey-Cheng commented:
Sorry for the mistake I've made. I'll correct this, update the description, and check the output for each one

Zoey-Cheng commented:
I've updated and tested the result again in gitlab MR, seems ok. Could you help review that?

reedwm commented on xla/tools/multihost_hlo_runner/hlo_runner_main.cc:
```diff
@@ -163,9 +164,24 @@ RunningOptionsFromFlags(const HloRunnerConfig& opts) {
   FunctionalHloRunner::RunningOptions out;
   TF_ASSIGN_OR_RETURN(out.module_argument_mode,
                       ArgumentModeFromString(opts.hlo_argument_mode));
+  // add parser for output mode (from functional_hlo_runner.h::AbslParseFlag)
```
Why not just call AbslParseFlag?

If there is a good reason, I'd remove this comment, since it doesn't add much value.

reedwm commented on xla/tools/multihost_hlo_runner/hlo_runner_main.cc:
```diff
@@ -419,7 +435,15 @@ int main(int argc, char** argv) {
+          "Possible values: return_outputs (default), not_return_outputs, "
+          "return_device0_outputs (return outputs only from logical device 0). "
+          "If outputs are not returned, outputs are still computed but the "
+          "potentially slow device-to-host copy is skipped. "),
```
Remove the space at the end of the comment

Zoey-Cheng replied:
Deleted. Also tested this through `--help`

Zoey-Cheng commented:
Addressed previous comments:
- Replaced the explicit parser with a call to the parser defined in `functional_hlo_runner.h`, and error handling with the error handling implemented in `functional_hlo_runner.cc`.
- Retested with various parameter values and verified the `--help` output.
Requesting another review. Thank you!

Zoey-Cheng commented on xla/tools/multihost_hlo_runner/hlo_runner_main.cc:
```diff
@@ -163,9 +164,17 @@ RunningOptionsFromFlags(const HloRunnerConfig& opts) {
                       ArgumentModeFromString(opts.hlo_argument_mode));
+  // add parser for output mode (matching from
+  // functional_hlo_runner.h::AbslParseFlag, error generate from
+  // functional_hlo_runner.cc)
```
Get. I deleted this
